File: shared-config/generators/generate-env.py
#!/usr/bin/env python3
"""
ERP Suite Environment Generator for Python modules
Generates environment files from YAML configurations for Python-based services
"""

import argparse
import os
import sys
import yaml
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Optional


class ConfigLoader:
    """Loads and processes configuration files"""
    
    def __init__(self, workspace_root: Path):
        self.workspace_root = workspace_root
        self.config_dir = workspace_root / "erp-suite" / "shared-config"
    
    def load_environment_config(self, environment: str) -> Dict[str, Any]:
        """Load configuration for a specific environment"""
        config_file = self.config_dir / "environments" / f"{environment}.yaml"
        
        if not config_file.exists():
            raise FileNotFoundError(f"Configuration file not found: {config_file}")
        
        with open(config_file, 'r') as f:
            content = f.read()
        
        # Expand environment variables
        expanded_content = os.path.expandvars(content)
        
        return yaml.safe_load(expanded_content)


class EnvironmentGenerator:
    """Generates environment files for Python modules"""
    
    def __init__(self, config: Dict[str, Any]):
        self.config = config
    
    def generate_env_content(self, module: str, environment: str) -> str:
        """Generate environment file content for a Python module"""
        
        content = []
        
        # Header
        content.extend([
            f"# Generated environment file for {module} module",
            f"# Environment: {environment}",
            f"# Generated at: {datetime.now().isoformat()}",
            "",
            "# ============================================================================",
            "# ENVIRONMENT",
            "# ============================================================================",
            f"ENVIRONMENT={self.config['environment']['name']}",
            f"DEBUG={str(self.config['environment']['debug']).lower()}",
            f"LOG_LEVEL={self.config['environment']['log_level']}",
            f"HOT_RELOAD={str(self.config['environment']['hot_reload']).lower()}",
            "",
        ])
        
        # Database connections
        content.extend([
            "# ============================================================================",
            "# DATABASE CONNECTIONS",
            "# ============================================================================",
            "",
        ])
        
        # PostgreSQL
        pg_config = self.config['databases']['postgresql']
        content.extend([
            "# PostgreSQL",
            f"DB_HOST={pg_config['host']}",
            f"DB_PORT={pg_config['port']}",
            f"DB_USER={pg_config['username']}",
            f"DB_PASSWORD={pg_config['password']}",
            f"DB_SSL_MODE={pg_config['ssl_mode']}",
            f"DB_MAX_CONNECTIONS={pg_config['max_connections']}",
            f"DB_CONNECTION_TIMEOUT={pg_config['connection_timeout']}",
        ])
        
        # Module-specific database
        if module in pg_config['databases']:
            db_name = pg_config['databases'][module]
            content.append(f"DB_NAME={db_name}")
            db_url = f"postgresql://{pg_config['username']}:{pg_config['password']}@{pg_config['host']}:{pg_config['port']}/{db_name}?sslmode={pg_config['ssl_mode']}"
            content.append(f"DATABASE_URL={db_url}")
        
        content.append("")
        
        # MongoDB
        mongo_config = self.config['databases']['mongodb']
        content.extend([
            "# MongoDB",
            f"MONGODB_HOST={mongo_config['host']}",
            f"MONGODB_PORT={mongo_config['port']}",
            f"MONGODB_USER={mongo_config['username']}",
            f"MONGODB_PASSWORD={mongo_config['password']}",
            f"MONGODB_AUTH_SOURCE={mongo_config['auth_source']}",
            f"MONGODB_MAX_POOL_SIZE={mongo_config['max_pool_size']}",
        ])
        
        # MongoDB URL
        mongo_db = mongo_config['databases'].get('analytics', 'erp_analytics')
        mongo_url = f"mongodb://{mongo_config['username']}:{mongo_config['password']}@{mongo_config['host']}:{mongo_config['port']}/{mongo_db}?authSource={mongo_config['auth_source']}"
        content.append(f"MONGODB_URL={mongo_url}")
        content.append("")
        
        # Redis
        redis_config = self.config['databases']['redis']
        content.extend([
            "# Redis",
            f"REDIS_HOST={redis_config['host']}",
            f"REDIS_PORT={redis_config['port']}",
            f"REDIS_PASSWORD={redis_config['password']}",
            f"REDIS_MAX_CONNECTIONS={redis_config['max_connections']}",
        ])
        
        # Redis URL
        redis_db = redis_config['databases'].get('cache', 0)
        redis_url = f"redis://:{redis_config['password']}@{redis_config['host']}:{redis_config['port']}/{redis_db}"
        content.append(f"REDIS_URL={redis_url}")
        content.append("")
        
        # Qdrant (for AI module)
        if module == 'ai':
            qdrant_config = self.config['databases']['qdrant']
            content.extend([
                "# Qdrant Vector Database",
                f"QDRANT_HOST={qdrant_config['host']}",
                f"QDRANT_HTTP_PORT={qdrant_config['http_port']}",
                f"QDRANT_GRPC_PORT={qdrant_config['grpc_port']}",
            ])
            
            if qdrant_config.get('api_key'):
                content.append(f"QDRANT_API_KEY={qdrant_config['api_key']}")
            
            ssl_prefix = "https" if qdrant_config.get('ssl') else "http"
            qdrant_url = f"{ssl_prefix}://{qdrant_config['host']}:{qdrant_config['http_port']}"
            content.append(f"QDRANT_URL={qdrant_url}")
            content.append("")
        
        # Kafka
        kafka_config = self.config['messaging']['kafka']
        content.extend([
            "# ============================================================================",
            "# MESSAGE BROKERS",
            "# ============================================================================",
            "",
            "# Kafka",
            f"KAFKA_BROKERS={','.join(kafka_config['brokers'])}",
            f"KAFKA_SECURITY_PROTOCOL={kafka_config['security_protocol']}",
        ])
        
        if kafka_config.get('sasl_mechanism'):
            content.append(f"KAFKA_SASL_MECHANISM={kafka_config['sasl_mechanism']}")
        if kafka_config.get('sasl_username'):
            content.append(f"KAFKA_SASL_USERNAME={kafka_config['sasl_username']}")
        if kafka_config.get('sasl_password'):
            content.append(f"KAFKA_SASL_PASSWORD={kafka_config['sasl_password']}")
        
        # Kafka topics
        content.extend([
            "",
            "# Kafka Topics",
            f"KAFKA_TOPIC_AUTH={kafka_config['topics']['auth_events']}",
            f"KAFKA_TOPIC_USER={kafka_config['topics']['user_events']}",
            f"KAFKA_TOPIC_BUSINESS={kafka_config['topics']['business_events']}",
            f"KAFKA_TOPIC_SYSTEM={kafka_config['topics']['system_events']}",
            f"KAFKA_TOPIC_AI={kafka_config['topics']['ai_events']}",
            f"KAFKA_TOPIC_NOTIFICATIONS={kafka_config['topics']['notification_events']}",
        ])
        
        # Module-specific consumer group
        service_key = f"{module}_service"
        if service_key in kafka_config['consumer_groups']:
            content.append(f"KAFKA_CONSUMER_GROUP={kafka_config['consumer_groups'][service_key]}")
        
        content.append("")
        
        # Elasticsearch
        es_config = self.config['search']['elasticsearch']
        content.extend([
            "# ============================================================================",
            "# SEARCH & ANALYTICS",
            "# ============================================================================",
            "",
            "# Elasticsearch",
            f"ELASTICSEARCH_HOST={es_config['host']}",
            f"ELASTICSEARCH_PORT={es_config['port']}",
            f"ELASTICSEARCH_USERNAME={es_config['username']}",
            f"ELASTICSEARCH_PASSWORD={es_config['password']}",
            f"ELASTICSEARCH_USE_SSL={str(es_config['use_ssl']).lower()}",
            f"ELASTICSEARCH_VERIFY_CERTS={str(es_config['verify_certs']).lower()}",
        ])
        
        # Elasticsearch URL
        es_protocol = "https" if es_config['use_ssl'] else "http"
        es_url = f"{es_protocol}://{es_config['username']}:{es_config['password']}@{es_config['host']}:{es_config['port']}"
        content.append(f"ELASTICSEARCH_URL={es_url}")
        content.append("")
        
        # Monitoring
        content.extend([
            "# ============================================================================",
            "# MONITORING & OBSERVABILITY",
            "# ============================================================================",
            "",
        ])
        
        # Prometheus
        prom_config = self.config['monitoring']['prometheus']
        content.extend([
            "# Prometheus",
            f"PROMETHEUS_HOST={prom_config['host']}",
            f"PROMETHEUS_PORT={prom_config['port']}",
            f"PROMETHEUS_URL=http://{prom_config['host']}:{prom_config['port']}",
            "",
        ])
        
        # Jaeger
        jaeger_config = self.config['monitoring']['jaeger']
        content.extend([
            "# Jaeger",
            f"JAEGER_HOST={jaeger_config['host']}",
            f"JAEGER_PORT={jaeger_config['port']}",
            f"JAEGER_GRPC_PORT={jaeger_config['grpc_port']}",
            f"JAEGER_HTTP_PORT={jaeger_config['http_port']}",
            f"JAEGER_AGENT_HOST={jaeger_config['agent_host']}",
            f"JAEGER_AGENT_PORT={jaeger_config['agent_port']}",
            f"JAEGER_ENDPOINT=http://{jaeger_config['host']}:{jaeger_config['http_port']}/api/traces",
            "",
        ])
        
        # Logging
        log_config = self.config['monitoring']['logging']
        content.extend([
            "# Logging",
            f"LOG_LEVEL={log_config['level']}",
            f"LOG_FORMAT={log_config['format']}",
            f"LOG_OUTPUT={log_config['output']}",
            "",
        ])
        
        # Security
        content.extend([
            "# ============================================================================",
            "# SECURITY",
            "# ============================================================================",
            "",
        ])
        
        # JWT
        jwt_config = self.config['security']['jwt']
        content.extend([
            "# JWT",
            f"JWT_SECRET={jwt_config['secret']}",
            f"JWT_ACCESS_EXPIRY={jwt_config['access_token_expiry']}",
            f"JWT_REFRESH_EXPIRY={jwt_config['refresh_token_expiry']}",
            f"JWT_ALGORITHM={jwt_config['algorithm']}",
            "",
        ])
        
        # CORS
        cors_config = self.config['security']['cors']
        content.extend([
            "# CORS",
            f"ALLOWED_ORIGINS={','.join(cors_config['allowed_origins'])}",
            f"ALLOWED_METHODS={','.join(cors_config['allowed_methods'])}",
            f"ALLOWED_HEADERS={','.join(cors_config['allowed_headers'])}",
            f"ALLOW_CREDENTIALS={str(cors_config['allow_credentials']).lower()}",
            "",
        ])
        
        # External integrations
        content.extend([
            "# ============================================================================",
            "# EXTERNAL INTEGRATIONS",
            "# ============================================================================",
            "",
        ])
        
        # AI (for AI module)
        if module == 'ai':
            ai_config = self.config['external']['ai']['openai']
            content.extend([
                "# OpenAI",
                f"OPENAI_API_KEY={ai_config['api_key']}",
                f"OPENAI_MODEL={ai_config['model']}",
                f"OPENAI_MAX_TOKENS={ai_config['max_tokens']}",
                "",
            ])
        
        # Service-specific configurations
        content.extend([
            "# ============================================================================",
            "# SERVICE-SPECIFIC CONFIGURATIONS",
            "# ============================================================================",
            "",
            f"# Module identification",
            f"MODULE_NAME={module}",
            f"SERVICE_NAME={module}-service",
        ])
        
        # Service ports
        service_key = f"{module}_service"
        if service_key in self.config['services']:
            service_config = self.config['services'][service_key]
            if 'http_port' in service_config:
                content.append(f"HTTP_PORT={service_config['http_port']}")
            if 'grpc_port' in service_config:
                content.append(f"GRPC_PORT={service_config['grpc_port']}")
            if 'port' in service_config:
                content.append(f"PORT={service_config['port']}")
        
        return '\n'.join(content) + '\n'


def find_workspace_root() -> Path:
    """Find the workspace root directory"""
    current = Path.cwd()
    
    while current != current.parent:
        if (current / '.git').exists() or \
           (current / 'go.mod').exists() or \
           (current / 'package.json').exists() or \
           (current / 'pyproject.toml').exists():
            return current
        current = current.parent
    
    raise FileNotFoundError("Workspace root not found")


def main():
    parser = argparse.ArgumentParser(
        description="Generate environment files for Python modules"
    )
    parser.add_argument(
        "--env", 
        default="development",
        choices=["development", "staging", "production", "testing"],
        help="Environment (default: development)"
    )
    parser.add_argument(
        "--module",
        default="ai",
        choices=["ai", "notification"],
        help="Module name (default: ai)"
    )
    parser.add_argument(
        "--output",
        help="Output file path (default: .env.{module}.{environment})"
    )
    parser.add_argument(
        "--verbose",
        action="store_true",
        help="Verbose output"
    )
    
    args = parser.parse_args()
    
    try:
        if args.verbose:
            print(f"Generating environment file for module '{args.module}' in environment '{args.env}'")
        
        # Find workspace root
        workspace_root = find_workspace_root()
        
        # Load configuration
        config_loader = ConfigLoader(workspace_root)
        config = config_loader.load_environment_config(args.env)
        
        # Generate environment content
        generator = EnvironmentGenerator(config)
        env_content = generator.generate_env_content(args.module, args.env)
        
        # Determine output file
        output_file = args.output or f".env.{args.module}.{args.env}"
        
        # Write output file
        with open(output_file, 'w') as f:
            f.write(env_content)
        
        if args.verbose:
            print(f"Successfully generated environment file: {output_file}")
        
        print(f"Environment file generated: {output_file}")
        
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: shared-config/generators/generate-env.js
#!/usr/bin/env node
/**
 * ERP Suite Environment Generator for Node.js modules
 * Generates environment files from YAML configurations for Node.js-based services
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const { program } = require('commander');

class ConfigLoader {
    constructor(workspaceRoot) {
        this.workspaceRoot = workspaceRoot;
        this.configDir = path.join(workspaceRoot, 'erp-suite', 'shared-config');
    }

    loadEnvironmentConfig(environment) {
        const configFile = path.join(this.configDir, 'environments', `${environment}.yaml`);
        
        if (!fs.existsSync(configFile)) {
            throw new Error(`Configuration file not found: ${configFile}`);
        }

        let content = fs.readFileSync(configFile, 'utf8');
        
        // Simple environment variable expansion
        content = content.replace(/\$\{([^}]+)\}/g, (match, varName) => {
            const [name, defaultValue] = varName.split(':');
            return process.env[name] || defaultValue || match;
        });

        return yaml.load(content);
    }
}

class EnvironmentGenerator {
    constructor(config) {
        this.config = config;
    }

    generateEnvContent(module, environment) {
        const content = [];

        // Header
        content.push(
            `# Generated environment file for ${module} module`,
            `# Environment: ${environment}`,
            `# Generated at: ${new Date().toISOString()}`,
            '',
            '# ============================================================================',
            '# ENVIRONMENT',
            '# ============================================================================',
            `ENVIRONMENT=${this.config.environment.name}`,
            `DEBUG=${this.config.environment.debug}`,
            `LOG_LEVEL=${this.config.environment.log_level}`,
            `HOT_RELOAD=${this.config.environment.hot_reload}`,
            ''
        );

        // Node.js specific
        content.push(
            '# Node.js Environment',
            `NODE_ENV=${environment === 'production' ? 'production' : 'development'}`,
            ''
        );

        // Database connections
        content.push(
            '# ============================================================================',
            '# DATABASE CONNECTIONS',
            '# ============================================================================',
            ''
        );

        // PostgreSQL
        const pgConfig = this.config.databases.postgresql;
        content.push(
            '# PostgreSQL',
            `DB_HOST=${pgConfig.host}`,
            `DB_PORT=${pgConfig.port}`,
            `DB_USER=${pgConfig.username}`,
            `DB_PASSWORD=${pgConfig.password}`,
            `DB_SSL_MODE=${pgConfig.ssl_mode}`,
            `DB_MAX_CONNECTIONS=${pgConfig.max_connections}`,
            `DB_CONNECTION_TIMEOUT=${pgConfig.connection_timeout}`
        );

        // Module-specific database
        if (pgConfig.databases[module]) {
            const dbName = pgConfig.databases[module];
            content.push(`DB_NAME=${dbName}`);
            const dbUrl = `postgresql://${pgConfig.username}:${pgConfig.password}@${pgConfig.host}:${pgConfig.port}/${dbName}?sslmode=${pgConfig.ssl_mode}`;
            content.push(`DATABASE_URL=${dbUrl}`);
        }

        content.push('');

        // MongoDB
        const mongoConfig = this.config.databases.mongodb;
        content.push(
            '# MongoDB',
            `MONGODB_HOST=${mongoConfig.host}`,
            `MONGODB_PORT=${mongoConfig.port}`,
            `MONGODB_USER=${mongoConfig.username}`,
            `MONGODB_PASSWORD=${mongoConfig.password}`,
            `MONGODB_AUTH_SOURCE=${mongoConfig.auth_source}`,
            `MONGODB_MAX_POOL_SIZE=${mongoConfig.max_pool_size}`
        );

        // MongoDB URL
        const mongoDb = mongoConfig.databases.analytics || 'erp_analytics';
        const mongoUrl = `mongodb://${mongoConfig.username}:${mongoConfig.password}@${mongoConfig.host}:${mongoConfig.port}/${mongoDb}?authSource=${mongoConfig.auth_source}`;
        content.push(`MONGODB_URL=${mongoUrl}`, '');

        // Redis
        const redisConfig = this.config.databases.redis;
        content.push(
            '# Redis',
            `REDIS_HOST=${redisConfig.host}`,
            `REDIS_PORT=${redisConfig.port}`,
            `REDIS_PASSWORD=${redisConfig.password}`,
            `REDIS_MAX_CONNECTIONS=${redisConfig.max_connections}`
        );

        // Redis URL
        const redisDb = redisConfig.databases.cache || 0;
        const redisUrl = `redis://:${redisConfig.password}@${redisConfig.host}:${redisConfig.port}/${redisDb}`;
        content.push(`REDIS_URL=${redisUrl}`, '');

        // WebSocket configuration (for frontend modules)
        if (['frontend', 'admin', 'websocket'].includes(module)) {
            const wsConfig = this.config.realtime.websocket;
            content.push(
                '# ============================================================================',
                '# REAL-TIME COMMUNICATION',
                '# ============================================================================',
                '',
                '# WebSocket',
                `WEBSOCKET_HOST=${wsConfig.host}`,
                `WEBSOCKET_PORT=${wsConfig.port}`,
                `WEBSOCKET_PATH=${wsConfig.path}`
            );

            const wsProtocol = wsConfig.ssl ? 'wss' : 'ws';
            const wsUrl = `${wsProtocol}://${wsConfig.host}:${wsConfig.port}`;
            content.push(
                `WEBSOCKET_URL=${wsUrl}`,
                `WEBSOCKET_CORS_ORIGINS=${wsConfig.cors_origins.join(',')}`,
                ''
            );
        }

        // Security
        content.push(
            '# ============================================================================',
            '# SECURITY',
            '# ============================================================================',
            ''
        );

        // JWT
        const jwtConfig = this.config.security.jwt;
        content.push(
            '# JWT',
            `JWT_SECRET=${jwtConfig.secret}`,
            `JWT_ACCESS_EXPIRY=${jwtConfig.access_token_expiry}`,
            `JWT_REFRESH_EXPIRY=${jwtConfig.refresh_token_expiry}`,
            `JWT_ALGORITHM=${jwtConfig.algorithm}`,
            ''
        );

        // CORS
        const corsConfig = this.config.security.cors;
        content.push(
            '# CORS',
            `ALLOWED_ORIGINS=${corsConfig.allowed_origins.join(',')}`,
            `ALLOWED_METHODS=${corsConfig.allowed_methods.join(',')}`,
            `ALLOWED_HEADERS=${corsConfig.allowed_headers.join(',')}`,
            `ALLOW_CREDENTIALS=${corsConfig.allow_credentials}`,
            ''
        );

        // Service URLs (for frontend modules)
        if (['frontend', 'admin'].includes(module)) {
            content.push(
                '# ============================================================================',
                '# SERVICE URLS',
                '# ============================================================================',
                ''
            );

            // API URLs
            const authService = this.config.services.auth_service;
            const crmService = this.config.services.crm_service;
            const hrmService = this.config.services.hrm_service;
            const financeService = this.config.services.finance_service;
            const inventoryService = this.config.services.inventory_service;
            const projectsService = this.config.services.projects_service;
            const aiService = this.config.services.ai_service;

            content.push(
                '# API Service URLs',
                `NEXT_PUBLIC_AUTH_API_URL=http://${authService.host}:${authService.http_port}`,
                `NEXT_PUBLIC_CRM_API_URL=http://${crmService.host}:${crmService.http_port}`,
                `NEXT_PUBLIC_HRM_API_URL=http://${hrmService.host}:${hrmService.http_port}`,
                `NEXT_PUBLIC_FINANCE_API_URL=http://${financeService.host}:${financeService.http_port}`,
                `NEXT_PUBLIC_INVENTORY_API_URL=http://${inventoryService.host}:${inventoryService.http_port}`,
                `NEXT_PUBLIC_PROJECTS_API_URL=http://${projectsService.host}:${projectsService.http_port}`,
                `NEXT_PUBLIC_AI_API_URL=http://${aiService.host}:${aiService.http_port}`,
                ''
            );

            // WebSocket URL for frontend
            const wsConfig = this.config.realtime.websocket;
            const wsProtocol = wsConfig.ssl ? 'wss' : 'ws';
            content.push(
                '# Real-time Communication',
                `NEXT_PUBLIC_WEBSOCKET_URL=${wsProtocol}://${wsConfig.host}:${wsConfig.port}`,
                ''
            );
        }

        // External integrations
        content.push(
            '# ============================================================================',
            '# EXTERNAL INTEGRATIONS',
            '# ============================================================================',
            ''
        );

        // Email
        const emailConfig = this.config.external.email;
        content.push(
            '# Email',
            `EMAIL_PROVIDER=${emailConfig.provider}`,
            `SMTP_HOST=${emailConfig.smtp_host}`,
            `SMTP_PORT=${emailConfig.smtp_port}`,
            `SMTP_USERNAME=${emailConfig.smtp_username}`,
            `SMTP_PASSWORD=${emailConfig.smtp_password}`,
            `EMAIL_FROM_ADDRESS=${emailConfig.from_address}`
        );

        if (emailConfig.use_tls !== undefined) {
            content.push(`SMTP_USE_TLS=${emailConfig.use_tls}`);
        }

        content.push('');

        // Storage
        const storageConfig = this.config.external.storage;
        content.push(
            '# Storage',
            `STORAGE_PROVIDER=${storageConfig.provider}`
        );

        if (storageConfig.local_path) {
            content.push(`STORAGE_LOCAL_PATH=${storageConfig.local_path}`);
        }

        if (storageConfig.s3_bucket) {
            content.push(
                `S3_BUCKET=${storageConfig.s3_bucket}`,
                `S3_REGION=${storageConfig.s3_region}`,
                `S3_ACCESS_KEY=${storageConfig.s3_access_key}`,
                `S3_SECRET_KEY=${storageConfig.s3_secret_key}`
            );
        }

        content.push('');

        // Payment (for frontend modules)
        if (['frontend', 'admin'].includes(module)) {
            const stripeConfig = this.config.external.payment.stripe;
            content.push(
                '# Payment (Stripe)',
                `NEXT_PUBLIC_STRIPE_PUBLISHABLE_KEY=${stripeConfig.publishable_key}`,
                `STRIPE_SECRET_KEY=${stripeConfig.secret_key}`,
                `STRIPE_WEBHOOK_SECRET=${stripeConfig.webhook_secret}`,
                ''
            );
        }

        // Feature flags (for frontend modules)
        if (['frontend', 'admin'].includes(module)) {
            content.push(
                '# ============================================================================',
                '# FEATURE FLAGS',
                '# ============================================================================',
                ''
            );

            Object.entries(this.config.feature_flags).forEach(([flag, enabled]) => {
                content.push(`NEXT_PUBLIC_FEATURE_${flag.toUpperCase()}=${enabled}`);
            });

            content.push('');
        }

        // Module-specific configurations
        content.push(
            '# ============================================================================',
            '# MODULE-SPECIFIC CONFIGURATIONS',
            '# ============================================================================',
            '',
            '# Module identification',
            `MODULE_NAME=${module}`,
            `SERVICE_NAME=${module}-service`
        );

        // Service ports
        const serviceKey = `${module}_service`;
        if (this.config.services[serviceKey]) {
            const serviceConfig = this.config.services[serviceKey];
            if (serviceConfig.http_port) {
                content.push(`HTTP_PORT=${serviceConfig.http_port}`);
            }
            if (serviceConfig.grpc_port) {
                content.push(`GRPC_PORT=${serviceConfig.grpc_port}`);
            }
            if (serviceConfig.port) {
                content.push(`PORT=${serviceConfig.port}`);
            }
        } else if (this.config.services[module]) {
            const serviceConfig = this.config.services[module];
            if (serviceConfig.port) {
                content.push(`PORT=${serviceConfig.port}`);
            }
        }

        // Frontend-specific configurations
        if (['frontend', 'admin'].includes(module)) {
            content.push(
                '',
                '# Next.js specific',
                `NEXTAUTH_URL=http://localhost:${this.config.services[module]?.port || 3000}`,
                `NEXTAUTH_SECRET=${jwtConfig.secret}`
            );
        }

        return content.join('\n') + '\n';
    }
}

function findWorkspaceRoot() {
    let current = process.cwd();
    
    while (current !== path.dirname(current)) {
        if (fs.existsSync(path.join(current, '.git')) ||
            fs.existsSync(path.join(current, 'go.mod')) ||
            fs.existsSync(path.join(current, 'package.json'))) {
            return current;
        }
        current = path.dirname(current);
    }
    
    throw new Error('Workspace root not found');
}

function main() {
    program
        .option('--env <environment>', 'Environment (development, staging, production, testing)', 'development')
        .option('--module <module>', 'Module name (frontend, admin, websocket)', 'frontend')
        .option('--output <file>', 'Output file path (default: .env.{module}.{environment})')
        .option('--verbose', 'Verbose output')
        .parse();

    const options = program.opts();

    try {
        if (options.verbose) {
            console.log(`Generating environment file for module '${options.module}' in environment '${options.env}'`);
        }

        // Find workspace root
        const workspaceRoot = findWorkspaceRoot();

        // Load configuration
        const configLoader = new ConfigLoader(workspaceRoot);
        const config = configLoader.loadEnvironmentConfig(options.env);

        // Generate environment content
        const generator = new EnvironmentGenerator(config);
        const envContent = generator.generateEnvContent(options.module, options.env);

        // Determine output file
        const outputFile = options.output || `.env.${options.module}.${options.env}`;

        // Write output file
        fs.writeFileSync(outputFile, envContent);

        if (options.verbose) {
            console.log(`Successfully generated environment file: ${outputFile}`);
        }

        console.log(`Environment file generated: ${outputFile}`);

    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

// Check if required dependencies are available
try {
    require('js-yaml');
    require('commander');
} catch (error) {
    console.error('Missing required dependencies. Please install them with:');
    console.error('npm install js-yaml commander');
    process.exit(1);
}

if (require.main === module) {
    main();
}

module.exports = { ConfigLoader, EnvironmentGenerator, findWorkspaceRoot };